ona/tasks.py

```diff
@@ -0,0 +1,69 @@
+from celery.task import Task
+from celery.utils.log import get_task_logger
+from celery.exceptions import SoftTimeLimitExceeded
+from django.core.exceptions import ObjectDoesNotExist
+import requests
+from requests.exceptions import HTTPError
+from .models import Submission
+
+logger = get_task_logger(__name__)
+
+
+class Send_Submission(Task):
+
+    """
+    Task to load and construct submission and send it off
+    """
+    name = "ona.tasks.send_submission"
+
+    class FailedEventRequest(Exception):
+
+        """
+        The attempted task failed because of a non-200 HTTP return code.
+        """
+
+    def run(self, submission_id, **kwargs):
+        try:
+            submission = Submission.objects.get(pk=submission_id)
+            if submission.submitted is False:
+                integration = submission.integration.details
+                data = '{"submission": '+submission.content+', "id": \
+                    "'+integration["form_id"]+'"}'
+                try:
+                    r = requests.post(
+                        integration["url"],
+                        data=data,
+                        auth=(integration["username"],
+                              integration["password"]),
+                        headers={
+                            "Content-Type": "application/json",
+                        }
+                    )
+
+                    # Mark the submission as submitted
+                    submission.submitted = True
+                    submission.save()
+                    # Log the Ona response on the report
+                    response = r.json()
+                    report = submission.report
+                    if "error" in response:
+                        report.metadata["ona_response"] = response["error"]
+                    else:
+                        report.metadata["ona_response"] = response["message"]
+                    report.save()
+                except HTTPError as e:
+                    # retry message sending if in 500 range (3 default retries)
+                    if 500 < e.response.status_code < 599:
+                        raise self.retry(exc=e)
+                    else:
+                        raise e
+        except ObjectDoesNotExist:
+            logger.error('Missing Submission object', exc_info=True)
```

Comment:
This catch should only be around the `submission = Submission.objects.get(...)` line (otherwise it's too hard to be sure that this is the only line that could trigger this exception path.